### PR TITLE
Shuffling added in training.

### PR DESCRIPTION
**Pull request by fartashf**
Addressing #232. Running `mnist_blackbox.py` and `mnist_tutorial_tf.py` the numbers were in the range of .5%.

### CODE REVIEW

**Comment by goodfeli** [comment]
Can you be more specific about what the numbers were before and after?
.5% is a very large relative difference when the baseline error rate is 1%.

**Comment by npapernot** [comment]
Travis is reporting the following test fail: 

```
======================================================================
ERROR: test_mnist_tutorial_tf (test_mnist_tutorial_tf.TestMNISTTutorialTF)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/tensorflow/cleverhans/tests_tf/test_mnist_tutorial_tf.py", line 21, in test_mnist_tutorial_tf
    report = mnist_tutorial_tf.mnist_tutorial(**test_dataset_indices)
  File "/home/travis/build/tensorflow/cleverhans/tutorials/mnist_tutorial_tf.py", line 107, in mnist_tutorial
    args=train_params)
  File "/home/travis/build/tensorflow/cleverhans/cleverhans/utils_tf.py", line 166, in model_train
    shuffle(index_shuf)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.5/random.py", line 281, in shuffle
    x[i], x[j] = x[j], x[i]
TypeError: 'range' object does not support item assignment
-------------------- >> begin captured stdout << ---------------------
Extracting /tmp/train-images-idx3-ubyte.gz
Extracting /tmp/train-labels-idx1-ubyte.gz
Extracting /tmp/t10k-images-idx3-ubyte.gz
Extracting /tmp/t10k-labels-idx1-ubyte.gz
X_train shape: (5000, 28, 28, 1)
X_test shape: (333, 28, 28, 1)
--------------------- >> end captured stdout << ----------------------
```

**Comment by fartashf** [comment]
To be specific about the change in accuracy, I ran the scripts `mnist_blackbox.py` and `mnist_tutorial_tf.py` 4 times before and after the change. The change to the mean of accuracy is within one standard deviation.

`mnist_blackbox.py`
Legitimate examples:
Before: 99.08+/-0.10
After: 99.10+/-0.10

Adversarial examples:
Before: 73.89+/-3.21
After: 73.68+/-4.6

`mnist_tutorial_tf.py`
Legitimate examples:
Before: 99.20+/-0.05
After: 99.24+/-0.03

Adversarial examples:
Before: 98.56+/-0.21
After: 98.49+/-0.24

**Comment by fartashf** [comment]
The scripts were called with `nb_epochs=50`.

**Comment by goodfeli** [comment]
For `mnist_tutorial_tf.py`, are all of these numbers with adversarial training? What happens without adversarial training?

**Comment by fartashf** [comment]
Yes. The previous numbers were with adversarial training.

Also without adversarial training the numbers are within one standard
deviation.

Legitimate examples:
Before: 99.26+/-0.02
After: 99.24+/-0.03

Adversarial examples:
Before: 15.24+/-1.27
After: 15.5+/-1.44

File: cleverhans/utils_tf.py [hunk 1]
```diff
@@ -160,6 +161,10 @@ def model_train(sess, x, y, predictions, X_train, Y_train, save=False,
             nb_batches = int(math.ceil(float(len(X_train)) / args.batch_size))
             assert nb_batches * args.batch_size >= len(X_train)
 
+            # Indices to shuffle training set
+            index_shuf = list(range(len(X_train)))
+            shuffle(index_shuf)
```
Comment (goodfeli) [inline cleverhans/utils_tf.py]: Can you provide a way of seeding this? Otherwise it will make the tutorial tests more random

**Comment by fartashf** [comment]
Similar to the seeding of TenosorFlow, I set a fixed seed in the tutorials for the `random` module. This is instead of passing the seed as an argument to `model_train`.

**Comment by npapernot** [comment]
LGTM

**Comment by goodfeli** [comment]
(Since this didn't result in a consistent improvement, let's not merge until after the 2.0 tag in order to respect the code freeze)

**Pull request merged**